actgen/change_q.py

```diff
@@ -0,0 +1,242 @@
+        parser.add_argument('--delta_update', '-u', type=float, default=10.0,
+                            help='increase the q value by this much for every update applied')
+        parser.add_argument('--change_percentage_thresh', '-p', type=float, default=0.10,
+                            help='only changes past this percentage are considered when computing the metrics')
+        args, unknown = parser.parse_known_args()
+        other_args = {
+            (remove_prefix(key, '--'), val)
+            for (key, val) in zip(unknown[::2], unknown[1::2])
+        }
+        args.other_args = other_args
+        # yapf: enable
+        return args
+
+    def load_hyperparams(self, args):
+        params = utils.load_hyperparams(args.hyperparams)
+        for arg_name, arg_value in vars(args).items():
+            if arg_name == 'hyperparams':
+                continue
+            params[arg_name] = arg_value
+        for arg_name, arg_value in args.other_args:
+            utils.update_param(params, arg_name, arg_value)
+        return params
+
+    def setup(self):
+        seeding.seed(0, random, torch, np)
+        test_env = gym.make(self.params['env_name'])
+        test_env = wrap.FixedDurationHack(test_env)
+        test_env = wrap.DuplicateActions(test_env, self.params['duplicate'])
+        test_env = wrap.TorchInterface(test_env)
+        seeding.seed(1000 + self.params['seed'], gym, test_env)
+        self.test_env = test_env
+
+        assert self.params['agent'] == 'dqn'
+        self.agent = DQNAgent(test_env.observation_space, test_env.action_space, self.params)
+        # load saved model
+        if not self.params['test']:
+            self.agent.q.load(self.params['saved_model'])
+
+    def teardown(self):
+        pass
+
+    def original_q_values(self, state):
+        """
+        get the q values for all actions for a particular state
+        """
+        self.agent.q.eval()
+        q_values = self.agent.q(state.float())
+        return q_values.detach().numpy()[0]
+
+    def dqn_directed_update(self, state, action, toward_target):
+        """
+        update the dqn agent towards a target for one state.
+        specifically, update update q(state, action) towards toward_target.
+        returns the old q(state,a) and the updated q(state, a) for all a.
+        """
+        self.agent.q.train()
+        self.agent.optimizer.zero_grad()
+
+        old_q_values = self.agent.q(state.float())  # q(state, a) for all a
+        q_target = old_q_values.clone()
+        q_target[0][action] = toward_target
+
+        loss = torch.nn.functional.smooth_l1_loss(input=old_q_values, target=q_target)
+        loss.backward()
+        self.agent.optimizer.step()
+
+        self.agent.q.eval()
+        new_q_values = self.agent.q(state.float())
+
+        return old_q_values.detach().numpy()[0], new_q_values.detach().numpy()[0]
+
+    def plot(self, old_values, new_values, action_idx):
```

Comment:
This could probably be a static method. It doesn't need `self`, since `old_values` and `new_values` are both length `n` already

Reply:
ah good point!